keel-core/src/main/kotlin/com/netflix/spinnaker/keel/services/ApplicationService.kt

```diff
@@ -192,29 +201,38 @@ class ApplicationService(
    * This function assumes there's a single delivery config associated with the application.
    */
   fun getArtifactSummariesFor(application: String, limit: Int = DEFAULT_MAX_ARTIFACT_VERSIONS): List<ArtifactSummary> {
+    val environmentSummaries = getEnvironmentSummariesFor(application)
+    return getArtifactSummariesFor(application, environmentSummaries, limit)
+  }
+
+  /**
+   * If we've already calculated the env summaries, pass them in so we don't have to query again.
+   * It's non-trivial to pull that data.
+   */
+  fun getArtifactSummariesFor(application: String, envSummaries: List<EnvironmentSummary>, limit: Int = DEFAULT_MAX_ARTIFACT_VERSIONS): List<ArtifactSummary> {
     val deliveryConfig = try {
       repository.getDeliveryConfigForApplication(application)
     } catch (e: NoSuchDeliveryConfigException) {
       return emptyList()
     }
 
-    val environmentSummaries = getEnvironmentSummariesFor(application)
-
     val artifactSummaries = deliveryConfig.artifacts.map { artifact ->
       val artifactVersionSummaries = repository.artifactVersions(artifact, limit).map { artifactVersion ->
         val artifactSummariesInEnvironments = mutableSetOf<ArtifactSummaryInEnvironment>()
 
-        environmentSummaries.forEach { environmentSummary ->
+        envSummaries.forEach { environmentSummary ->
           val environment = deliveryConfig.environments.find { it.name == environmentSummary.name }!!
           environmentSummary.getArtifactPromotionStatus(artifact, artifactVersion.version)
             ?.let { status ->
-              buildArtifactSummaryInEnvironment(deliveryConfig, environment.name, artifact, artifactVersion.version, status)
-                ?.also {
-                  artifactSummariesInEnvironments.add(
-                    it.addStatefulConstraintSummaries(deliveryConfig, environment, artifactVersion.version)
-                      .addStatelessConstraintSummaries(deliveryConfig, environment, artifactVersion.version, artifact)
-                  )
-                }
+              if ( artifact.isUsedIn(environment)) { // only add a summary if the artifact is used in the environment
+                buildArtifactSummaryInEnvironment(deliveryConfig, environment.name, artifact, artifactVersion.version, status)
+                  ?.also {
+                    artifactSummariesInEnvironments.add(
+                      it.addStatefulConstraintSummaries(deliveryConfig, environment, artifactVersion.version)
+                        .addStatelessConstraintSummaries(deliveryConfig, environment, artifactVersion.version, artifact)
+                    )
+                  }
+              }
```

Comment:
^ is the big fix, specifically line 227.

Reply:
Now, it's true that we're still calculating the summary in the SQL. We're just filtering it out here, and this is the view that the UI uses. We can surely optimize this later if needed but focusing on stopping the really annoying user facing behavior for now